dynamic_instrumentation/src/instrument.rs

```diff
@@ -123,6 +129,198 @@ fn is_region_or_unsafe_ptr(ty: Ty) -> bool {
     ty.is_unsafe_ptr() || ty.is_region_ptr()
 }
 
+impl<'tcx> Visitor<'tcx> for CheckAddressTakenLocals<'_, 'tcx> {
+    /// Checks the right hand side of each assignment statement for taking the
+    /// address of a local. This may happen explicitly with something like
+    /// std::ptr::addr_of! or implicitly by taking a reference & of a local.
+    /// Instances/locations of address-taking are saved for a later walk over the MIR
+    /// AST.
+    fn visit_assign(&mut self, place: &Place<'tcx>, rvalue: &Rvalue<'tcx>, location: Location) {
+        let value_ty = rvalue.ty(self, self.tcx());
+
+        self.super_assign(place, rvalue, location);
+
+        match rvalue {
+            _ if !is_region_or_unsafe_ptr(value_ty) => {}
+
+            Rvalue::AddressOf(_, p) | Rvalue::Ref(_, _, p) if p.projection.is_empty() => {
+                self.address_taken.insert(p.local);
+            }
+            _ => (),
+        }
+    }
+}
+
+impl<'tcx> MutVisitor<'tcx> for RewriteAddressTakenLocals<'tcx> {
+    /// Rewrites an address-taken local in terms of its underlying address.
+    fn visit_place(
+        &mut self,
+        mut place: &mut Place<'tcx>,
+        context: PlaceContext,
+        location: Location,
+    ) {
+        // If we have found an address-taken local `_x`, substitute with `(*_y)` where `_y`
+        // is the address of `_x`.
+        if let Some(substitute) = self.local_substitute.get(&place.local) {
+            if context.is_use()
+                && (!context.is_place_assignment()
+                    || (context.is_place_assignment()
+                        && !place.is_indirect()
+                        && place.projection.len() > 0))
+                && !context.is_drop()
+                && !matches!(
+                    context,
+                    PlaceContext::NonMutatingUse(NonMutatingUseContext::Inspect)
+                )
+            {
+                // add deref
+                let projection = {
+                    let mut v = Vec::with_capacity(place.projection.len() + 1);
+                    v.extend([ProjectionElem::Deref]);
+                    v.extend(place.projection);
+                    self.tcx().intern_place_elems(&v)
+                };
+                // replace `_x` with `_y`
+                place.local = *substitute;
+                place.projection = projection;
+            }
+        }
+
+        self.super_place(place, context, location)
+    }
+
+    fn tcx<'a>(&'a self) -> TyCtxt<'tcx> {
+        self.tcx()
+    }
+
+    fn visit_body(&mut self, body: &mut Body<'tcx>) {
+        let mut local_address_statements = Vec::new();
```

Comment:
I think it would be helpful to use a `struct` with named fields instead of the tuple `(Place, BasicBlock, usize)`, which is confusing, especially for the `usize`, which is a statement index/id from what I can tell.

Reply:
used `Location` to help with that understanding